### ISSUE TITLE
Textures not loading on mobile devices browsers (including, and after v: 4.0.2)

### ISSUE THREAD

**Issue by Snky** [openfl/lime#989]
I have created my website in Starling for Haxe.

This setup will work on all browsers:
lime 4.0.1
openfl 4.9.2
starling 1.8.3 - 1.8.10 (any version)

If I upgrade lime to 4.0.2 or higher (if I switch both dependencies to 5.x.x for example), my website will work on desktop, but not on my mobile device(s) (such as: iPhone 4S (iOS), TF101 (Android)).

Going by the version history of lime, I would of assumed 4.0.1 would of been the problematic one:
"OpenGL API and AssetLibrary fixes"
But it is indeed 4.0.2:
"Android fixes, network improvements"

I managed to work out that, on the mobile device(s) browser(s) (by displaying text), textures are not being accessed (which are stored on my site), on lime versions 4.0.2 and higher.

The texture loading code I can show for this initial post is as so:

	public static function getSprite(name_:String = null):DisplayObject {
		var n:String = name_;
		if (null == n) return null;
		
		var t:Vector<Texture> = am.getTextures(n);
		var l:UInt = t.length;
		
		if ( l == 0 ) {
			n = "sTextureMissing";
			t = am.getTextures(n);
			l = t.length;
		}
		
		if (l == 0) {
			return null;
		}
		else
		if (l == 1) {
			var spr:Image = new Image(t[0]);
			spr.name = name_;
			spr.smoothing = TextureSmoothing.NONE;
			return spr;
		} else {
			var mc:MovieClip = new MovieClip(t);
			mc.name = name_;
			mc.smoothing = TextureSmoothing.NONE;
			return mc;
		}
	}

The result is that 'l' (length variable) is equal to '0' for mobile device(s) browser(s),
but again, on desktop, the length would be '1', as it successfully found the assets on my site.

I hope I posted in the correct GitHub, I do not feel like this is a starling related issue, nor an openfl issue, as it only occurs when I switch from lime 4.0.1 to lime 4.0.2.

Thank you for your time

**Comment by jgranick** [comment]
Does this problem occur for you with the Starling demo? Thanks :smile:

**Comment by Snky** [comment]
I'll get back to you with that answer, does the Starling demo load assets via URL strings?
I've never ran the demo before.

**Comment by Snky** [comment]
Tried: "..\starling\1,8,10\samples\demo" with Lime 5.0.3 and OpenFL 5.1.2, it works fine in HTML5 export.
How relevant is this if it is using embedded assets?

**Comment by Snky** [comment]
Tried: "..\starling\1,8,10\samples\demo" with Lime 5.0.3 and OpenFL 5.1.2, it works fine in HTML5 export.
How relevant is this if it is using embedded assets? My issue is regarding loading assets stored on the site.
Works fine on desktop browsers, but not on mobile browsers (past lime version 4.0.1).

**Comment by jgranick** [comment]
The Starling sample should not be using embedded assets, they are separate files, preloaded at runtime before they are used.

How are you loading your assets? `BitmapData.loadFromFile`, `URLLoader`, something else?

**Comment by Snky** [comment]
Good evening, I am enqueuing assets from a remote URL via the AssetManager.

    //scene
    otherclass.loadAssets("sSocial", handleAtlasesReady);

    private function handleAtlasesReady(assets:AssetManager):Void {
        //.. assets are loaded by this point
    }

    //other class
    public static function loadAssets(args:Array<String>, onComplete:AssetManager->Void):Void {
        am.textureFormat = Context3DTextureFormat.BGRA_PACKED;
        am.verbose = Capabilities.isDebugger;
        for (str in args){
            #if js
                var enqueueArray:Array<Dynamic> = new Array<Dynamic>();
                enqueueArray.push("http://www.snkygames.com/textures/" + scaleFactor + "/spr/" + str + "_png.png");
                enqueueArray.push("http://www.snkygames.com/textures/" + scaleFactor + "/spr/" + str + "_png.xml");
                am.enqueue(enqueueArray);
            #else
                //...
            #end
        }
		
        am.loadQueue(function(ratio:Float):Void {
            if (ratio == 1.0) {
                onComplete(am);
            }
        });
    }

"am" is my variable reference to AssetManager. I used the wrong term when I said the demo used embedded assets, more accurately, I could say, it is not accessing files stored on a remote URL.

**Comment by Snky** [comment]
Good evening, I am enqueuing assets from a remote URL via the AssetManager.

    //scene
    otherclass.loadAssets("sSocial", handleAtlasesReady);

    private function handleAtlasesReady(assets:AssetManager):Void {
        //.. assets are loaded by this point
    }

    //other class
    public static function loadAssets(args:Array<String>, onComplete:AssetManager->Void):Void {
        am.textureFormat = Context3DTextureFormat.BGRA_PACKED;
        am.verbose = Capabilities.isDebugger;
        for (str in args){
            #if js
                var enqueueArray:Array<Dynamic> = new Array<Dynamic>();
                enqueueArray.push("http://www.snkygames.com/textures/" + scaleFactor + "/spr/" + str + "_png.png");
                enqueueArray.push("http://www.snkygames.com/textures/" + scaleFactor + "/spr/" + str + "_png.xml");
                am.enqueue(enqueueArray);
            #else
                //...
            #end
        }
		
        am.loadQueue(function(ratio:Float):Void {
            if (ratio == 1.0) {
                onComplete(am);
            }
        });
    }

"am" is my variable reference to AssetManager. I used the wrong term when I said the demo used embedded assets, more accurately, I could say, it is not accessing files stored on a remote location.

**Comment by Snky** [comment]
Is there any other information I can provide to help get a solution to this issue @jgranick? I would like to catch up to the latest Lime & OpenFL combination, I was hoping to report some issues, but I wanted to do some testing on the latest versions.

**Comment by Snky** [comment]
I'm copying files from Lime 5.3.x to 4.0.1 and from openfl 6.0.x to 4.9.2, disabled the version detection too, this is painful to do, but it's the only way I know how to try and find the problematic change that may be stopping me from being able to use the latest versions - I'll update my finding as I trial and error some more. My current hunch is, that it is related to lime/utils/DataPointer.hx

**Comment by Snky** [comment]
I had my friend test my app on his mobile device (LG G5), and it worked fine for them with Lime 5.5.0. 

I am still considering this as critical, I do not consider my iPhone 4S to be outdated yet, and it worked with Lime 4.0.1. Additional information: It seems to land on an IO Error when trying to load the assets.

I think it can now be declared as device specific. Not sure how many devices it would affect.

BAD RESULT:
https://dl.dropbox.com/s/89vbwvtn9pvjh4w/2017-09-17%2021.35.51.png

DESIRED RESULT:
https://i.imgur.com/fmCy1er.png

So, to put it simply, Lime 4.0.1 makes the HTML5 export compatible with all devices, 4.0.2 and beyond support a limited number of devices.

**Comment by jgranick** [comment]
I think this was the only change in loading on HTML5, between 4.0.1 and 4.0.2:

https://github.com/openfl/lime/commit/d593685dcb12ac7fee79cf41119bfa84b99b6eb7

**Comment by jgranick** [comment]
Does it make a difference if you use a relative URL for loading your assets? Perhaps there is a cross-origin permissions issue. Can you tell if the assets load, or if there is an error in the console? Have you tried connected developer tools to the phone to get more information?

**Comment by Snky** [comment]
"Do you use a custom HTML template?"
I tried using the default 'index.html' generated if that is what you mean. Same issue.

"Does it make a difference if you use a relative URL for loading your assets?"
I am not sure what you mean by the relative in this context, do you have an example?

"Can you tell if the assets load, or if there is an error in the console?"
At the moment, I have it so, if the assets did not load, the text will appear saying 'device not supported' (e.g. 'snkygames.com' on my iPhone 4S). Unfortunately, for this device, and any Lime version above 4.0.1, it does not seem to load any assets.

"Have you tried connected developer tools to the phone to get more information?"
No, I'm not sure how to do this, is there a simple guide to doing this? As I would need to test this through a mobile browser, the desktop browsers have no issues(this probably means it cannot be a cross-origin permission issue, I have tried "*" on Header set Access-Control-Allow-Origin <> to no avail.

**Comment by Snky** [comment]
"Do you use a custom HTML template?"
I tried using the default 'index.html' generated if that is what you mean. Same issue.

"Does it make a difference if you use a relative URL for loading your assets?"
I am not sure what you mean by the relative in this context, do you have an example?

"Can you tell if the assets load, or if there is an error in the console?"
At the moment, I have it so, if the assets did not load, the text will appear saying 'device not supported' (e.g. 'snkygames.com' on my iPhone 4S). Unfortunately, for this device, and any Lime version above 4.0.1, it does not seem to load any assets.

"Have you tried connected developer tools to the phone to get more information?"
No, I'm not sure how to do this, is there a simple guide to doing this? As I would need to test this through a mobile device browser, the desktop browsers have no issues(this probably means it cannot be a cross-origin permission issue, I have tried "*" on Header set Access-Control-Allow-Origin <> to no avail).

**Comment by jgranick** [comment]
By relative URL, I mean "../path/to/asset.png" rather than "http://www.mysite.com/path/to/asset.png", sometimes it helps

As for debugging, here's a couple links :)

http://washamdev.com/2016/02/debug-a-website-in-ios-safari-on-windows/
https://moduscreate.com/enable-remote-web-inspector-in-ios-6/

**Comment by Snky** [comment]
I tried the relative path, thank you for that, good to know, it has helped me with some (off-topic) debugging.

I tried that first link, and I got it tracing errors!

Error: "Cross-origin image load denied by Cross-Origin Resource Sharing policy."
(Yikes, this was one of the first things I made sure to check.)

Any reason why I might be seeing that on > Lime 4.0.1, but not on 4.0.1?

**Comment by jgranick** [comment]
I'm not sure, the code should be similar:

https://github.com/openfl/lime/blob/e375e24959937d810ace05c84f1d44868a3bf882/lime/graphics/Image.hx#L857-L883

https://github.com/openfl/lime/blob/d593685dcb12ac7fee79cf41119bfa84b99b6eb7/lime/_backend/html5/HTML5HTTPRequest.hx#L302-L332

Maybe the path/uri value is slightly different between the versions?

**Comment by Snky** [comment]
Is that 'path/uri' something on my end? I can literally roll back to 4.0.1 and the most compatible OpenFL version and it will load fine with no issues without changing my code(wondering what else I can test now).

**Comment by Snky** [comment]
I commented out: 

     image.crossOrigin = "Anonymous";

In both (lime/lime/_backend/html5/)HTML5HTTPRequest.hx and (lime/lime/graphics/)Image.hx
and it works. Well.. maybe you'll know why that is, I don't, but I'm happy doing that edit every update.

**Comment by jgranick** [comment]
Oh, I think we have a workaround in the latest version:

https://github.com/openfl/lime/blob/9afbdbe586ef23166c1a97f2c65dca4b54d787f2/lime/_backend/html5/HTML5HTTPRequest.hx#L426-L430

If you are able, would you mind trying the latest version, and if it still does not work, can you try and add some traces inside the `__isSameOrigin` method, and perhaps we can improve it to automatically disable the `crossOrigin` field automatically? :smile:

**Comment by Snky** [comment]
I have the latest version as of earlier today, before my first message here(today). I can try some traces, what exactly will I be looking to capture?

**Comment by jgranick** [comment]
See here:

https://github.com/openfl/lime/blob/9afbdbe586ef23166c1a97f2c65dca4b54d787f2/lime/_backend/html5/HTML5HTTPRequest.hx#L362-L366

something like:

```haxe
trace ("-------");

trace (a.hostname, hostname);
trace (a.protocol, protocol);
trace (a.port, port);

trace (originHostname);
trace (originProtocol);
trace (originPort);

trace (sameOrigin);
```

**Comment by Snky** [comment]
HTML5HTTPRequest.hx:357: -------
HTML5HTTPRequest.hx:359: snkygames.com,snkygames.com
"HTML5HTTPRequest.hx:360: http:,http:"
HTML5HTTPRequest.hx:361: ,80
HTML5HTTPRequest.hx:363: snkygames.com
HTML5HTTPRequest.hx:364: http:
HTML5HTTPRequest.hx:365: 80
HTML5HTTPRequest.hx:367: true

Starling.hx:462: [Starling] Initialization complete.
Starling.hx:463: [Starling] Display Driver: OpenGL Vendor=WebKit Version=WebGL 1.0 (OpenGL ES 2.0 IMGSGX543-124.1) Renderer=WebKit WebGL GLSL=WebGL GLSL ES 1.0 (OpenGL ES GLSL ES 1.00)

HTML5HTTPRequest.hx:357: -------
HTML5HTTPRequest.hx:359: ,
"HTML5HTTPRequest.hx:360: data:,data:"
HTML5HTTPRequest.hx:361: ,
HTML5HTTPRequest.hx:363: snkygames.com
HTML5HTTPRequest.hx:364: http:
HTML5HTTPRequest.hx:365: 80
HTML5HTTPRequest.hx:367: false

A slight variance after Starling was initialised.

**Comment by jgranick** [comment]
maybe this would fix it:

```haxe
if (hostname == "") hostname = originHostname;
if (port == "") port = originPort;
```

**Comment by Snky** [comment]
I put those lines before the return:

    if (hostname == "") hostname = originHostname;
    if (port == "") port = originPort;
    return sameOrigin;

Near line 370~, seems that Cross-Origin error occurred again.

Starling.hx:462: [Starling] Initialization complete.
Starling.hx:463: [Starling] Display Driver: OpenGL Vendor=WebKit Version=WebGL 1.0 (OpenGL ES 2.0 IMGSGX543-124.1) Renderer=WebKit WebGL GLSL=WebGL GLSL ES 1.0 (OpenGL ES GLSL ES 1.00)
SnkyEngine.hx:383: AssetPack: 4
SnkyEngine.hx:383: AssetPack: 4
AssetManager.hx:1191: [AssetManager] Enqueuing 'sUtils_png'
SnkyEngine.hx:383: AssetPack: 4
HTML5HTTPRequest.hx:357: -------
HTML5HTTPRequest.hx:359: ,
"HTML5HTTPRequest.hx:360: data:,data:"
HTML5HTTPRequest.hx:361: ,
HTML5HTTPRequest.hx:363: snkygames.com
HTML5HTTPRequest.hx:364: http:
HTML5HTTPRequest.hx:365: 80
HTML5HTTPRequest.hx:367: false
Cross-origin image load denied by Cross-Origin Resource Sharing policy.

When doing this test again, I reset Image.hx back to its default(it's not an important class anymore right?).

**Comment by Snky** [comment]
I think I would need to add some form of:

    if (protocol == ...

To at least, pass this test, as that's a part of the sameOrigin query.

**Comment by Snky** [comment]
I think I would need to add some form of:

    if (protocol == ...

To at least, pass this test, as that's a part of the sameOrigin query. As going by my traces, after Starling is initialised, it's 'data:' not 'http:'.

**Comment by Snky** [comment]
I think I would need to add some form of:

    if (protocol == ...

To at least, pass this test, as that's a part of the sameOrigin query. 
Going by my traces, after Starling is initialised, the protocol has changed from 'http:' to 'data:'.

**Comment by Snky** [comment]
I think I would need to add some form of:

    if (protocol == ...

To at least, pass this test, as that's a part of the sameOrigin query. 
Going by my traces, after Starling is initialised, the protocol has changed from 'http:' to 'data:'.

Perhaps a solution would be to change the sameOrigin query on top of the addition of those two new lines mentioned above:

    //from
    var sameOrigin = (protocol != "file:") && hostname == originHostname && protocol == originProtocol && port == originPort;
    //to
    var sameOrigin = (protocol != "file:") && hostname == originHostname && port == originPort;

Technically, the protocol doesn't have anything to do with the location(origin).

**Comment by Snky** [comment]
On top of the addition of these two lines:

		if (hostname == "") hostname = originHostname;
		if (port == "") port = originPort;

Perhaps the solution would be to change the sameOrigin query:

    //from
		var sameOrigin = (protocol != "file:") && hostname == originHostname && protocol == originProtocol && port == originPort;
		//to
		var sameOrigin = (protocol != "file:") && hostname == originHostname && port == originPort;

Technically, the protocol doesn't have anything to do with the location(origin).

**Comment by Snky** [comment]
On top of the addition of these two lines:

		if (hostname == "") hostname = originHostname;
		if (port == "") port = originPort;

Perhaps the solution would be to change the sameOrigin query:

		//from
		var sameOrigin = (protocol != "file:") && hostname == originHostname && protocol == originProtocol && port == originPort;
		//to
		var sameOrigin = (protocol != "file:") && hostname == originHostname && port == originPort;

Technically, the protocol doesn't have anything to do with the location(origin).

**Comment by jgranick** [comment]
Okay, I think this commit should work to improve/fix the issue. Please give it a try, and let me know how it works :smile:

**Comment by Snky** [comment]
Works, thank you, now I can stay with the latest versions. 😄
